src/tool/cppwinrt/cppwinrt/helpers.h

```diff
@@ -246,17 +268,78 @@ namespace xlang
         return bases;
     }
 
+    static std::pair<uint16_t, uint16_t> get_version(TypeDef const& type)
+    {
+        uint32_t version{};
+
+        for (auto&& attribute : type.CustomAttribute())
+        {
+            auto name = attribute.TypeNamespaceAndName();
+
+            if (name.first != "Windows.Foundation.Metadata")
+            {
+                continue;
+            }
+
+            if (name.second == "ContractVersionAttribute")
+            {
+                version = get_attribute_value<uint32_t>(attribute, 1);
+                break;
+            }
+
+            if (name.second == "VersionAttribute")
+            {
+                version = get_attribute_value<uint32_t>(attribute, 0);
+                break;
+            }
+        }
+
+        return { HIWORD(version), LOWORD(version) };
+    }
+
     struct interface_info
     {
         TypeDef type;
         bool is_default{};
         bool defaulted{};
         bool overridable{};
         bool base{};
+        bool exclusive{};
+        bool fastabi{};
+        std::pair<uint16_t, uint16_t> version{};
         std::vector<std::vector<std::string>> generic_param_stack{};
     };
 
-    static void get_interfaces_impl(writer& w, std::map<std::string, interface_info>& result, bool defaulted, bool overridable, bool base, std::vector<std::vector<std::string>> const& generic_param_stack, std::pair<InterfaceImpl, InterfaceImpl>&& children)
+    using get_interfaces_t = std::vector<std::pair<std::string, interface_info>>;
+
```

Comment:
I *really* like this pattern.  consider using it more liberally (e.g., for get_version above)

Reply:
👍